### PR TITLE
Refurbish DependenciesPackager

### PR DESCRIPTION
**Pull request by troydai**
Split the Program into smaller components to improve maintainability and readability

/cc @pranavkm @javiercn

### CODE REVIEW

File: src/DependenciesPackager/CrossGenToolFileNames.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            ["debian.8"] = new CrossGenTool("crossgen", "libclrjit.so"),
+        };
+
+        private CrossGenTool(string crossgen, string clrjit)
+        {
+            CrossGen = crossgen;
+            ClrJit = clrjit;
+        }
+
+        public static CrossGenTool GetCrossGenTool(string runtimeMonitor)
```
Comment (javiercn) [inline src/DependenciesPackager/CrossGenToolFileNames.cs]: nit: This was probably me, but it should be runtimeMoniker not Monitor

File: src/DependenciesPackager/CrossGenToolFileNames.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            {
+                return result;
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        public static IEnumerable<string> FindAllCrossGen(string searchPath) =>
```
Comment (javiercn) [inline src/DependenciesPackager/CrossGenToolFileNames.cs]: Pranav points

File: src/DependenciesPackager/CrossgenContext.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                    {
+                        errorMessage.AppendLine(line);
+                    }
+                }
+            }
+
+            _logger.LogInformation(errorMessage.ToString());
+        }
+
+        public void CollectPackagies(string runtime, string restoreFolder)
```
Comment (javiercn) [inline src/DependenciesPackager/CrossgenContext.cs]: nit: Typo Packages

File: src/DependenciesPackager/CrossgenContext.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            var jitDir = Path.GetDirectoryName(jitExecutablePath);
+            foreach (var file in Directory.GetFiles(jitDir))
+            {
+                File.Copy(file, Path.Combine(_stagingFolder, Path.GetFileName(file)), overwrite: true);
+            }
+        }
+
+        public void Crossgen(string outputPath)
+        {
+            foreach (var pacakge in _packageEntries)
```
Comment (javiercn) [inline src/DependenciesPackager/CrossgenContext.cs]: nit: typo package

File: src/DependenciesPackager/CrossgenContext.cs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                package.Library.Identity.Name,
+                package.Library.Identity.Version.ToNormalizedString(),
+                Path.GetDirectoryName(asset.RelativePath));
+
+            _logger.LogInformation($"Creating sub directory on {subDirectoryPath}.");
+
+            return Directory.CreateDirectory(subDirectoryPath);
+        }
+
+        private string GetTempPublishFolder(string parent)
```
Comment (javiercn) [inline src/DependenciesPackager/CrossgenContext.cs]: Rename to GetTempStagingFolder

File: src/DependenciesPackager/IDisposableContext.cs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+﻿namespace DependenciesPackager
+{
+    internal interface IDisposableContext
```
Comment (javiercn) [inline src/DependenciesPackager/IDisposableContext.cs]: How is this different than IDisposable? Please remove and just use IDisposable

File: src/DependenciesPackager/OutputFilesContext.cs [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+                    Directory.Delete(dir, recursive: true);
+                }
+            }
+        }
+
+        public void PrintDiagnostics()
+        {
+            var files = Directory.GetFiles(OutputPath, "*", SearchOption.AllDirectories);
+            var notCrossgend = new List<string>();
+            var buf = new StringBuilder();
```
Comment (javiercn) [inline src/DependenciesPackager/OutputFilesContext.cs]: Do not abbreviate: buffer

File: src/DependenciesPackager/PackageRestoreContext.cs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            foreach (var file in filesToRemove)
+            {
+                _logger.LogInformation($"Removing file '{file}.");
+                File.Delete(file);
+            }
+        }
+
+        private int RunDotnetRestore()
+        {
+            var buf = new StringBuilder();
```
Comment (javiercn) [inline src/DependenciesPackager/PackageRestoreContext.cs]: unused

File: src/DependenciesPackager/ProcessRunner.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        public int Run()
+        {
+            if (_process != null)
+            {
+                throw new InvalidOperationException("The process has already been started.");
+            }
+
+            ProcessStartInfo processInfo = CreateProcessInfo();
```
Comment (javiercn) [inline src/DependenciesPackager/ProcessRunner.cs]: nit: var

File: src/DependenciesPackager/Program.cs [hunk 10]
```diff
@@ -98,20 +49,27 @@ private ILogger CreateLogger()
...
             _cliPath = cliPath;
-            _quiet = quiet;
-            _runtimes = runtimes;
+            _optRuntime = runtimes;
             _prefix = prefix;
             _keepTemporaryFiles = keepTemporaryFiles;
+            _quiet = quiet;
+
+            // set up logger
+            _logger = (new LoggerFactory())
```
Comment (javiercn) [inline src/DependenciesPackager/Program.cs]: extra parenthesis not needed

File: src/DependenciesPackager/ProjectContextExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+                {
+                    Library = dependency.Library,
+                    Assets = assets
+                });
+            }
+
+            return result;
+        }
+
+        public static string GetPublishFolderPath(this ProjectContext context, string publishFolder) =>
```
Comment (javiercn) [inline src/DependenciesPackager/ProjectContextExtensions.cs]: Should this be GetStagingFolderPath?

**Comment by javiercn** [comment]
:watch: Overall looks good. I would suggest that if you are performing this overhaul of the tool, you try and follow an alternative approach for doing crossgen that consists on passing the semi colon separated list of paths to the tool instead of copying everything to a folder and running crossgen. 
This will have a noticeable improvement in performance for the tool and if the size of the command line becomes a problem, you can write a response file in the following format

```
/nologo
/App_Paths path1;path2;....
```

And then call crossgen with crossgen.exe /in PathToin.dll /out /PathToOuthDll @response.rsp

This will avoid copying all the files to an intermediate folder which takes like 20% of the tools time if not more.

What I'm suggesting is, get all the dependencies from the project context, build a rsp file with /App_Paths <<fullpathlib1.dll;fullpathlib2.dll> and pass that to the crossgen input with @myresponsefile.rsp instead of copying the assemblies to a staging folder

**Comment by troydai** [comment]
PR updated for review comments.

Issue filed to track the improvement suggestion.
https://github.com/aspnet/DnxTools/issues/96

However I won't pursue the change right now. The refurbish took me a full day, and the next important thing is to address #94 . Should I have time or performance become a bigger issue then #96 will be picked up.

**Comment by troydai** [comment]
Ping @pranavkm @javiercn

**Comment by javiercn** [comment]
:shipit:

**Pull request merged**